// @inversifyjs/core/lib/index.js
import { getBindingId } from './binding/actions/getBindingId.js';
import { bindingScopeValues, } from './binding/models/BindingScope.js';
import { bindingTypeValues, } from './binding/models/BindingType.js';
import { ActivationsService, } from './binding/services/ActivationsService.js';
import { BindingService } from './binding/services/BindingService.js';
import { DeactivationsService, } from './binding/services/DeactivationsService.js';
import { decorate } from './decorator/actions/decorate.js';
import { InversifyCoreError } from './error/models/InversifyCoreError.js';
import { InversifyCoreErrorKind } from './error/models/InversifyCoreErrorKind.js';
import { getClassMetadata } from './metadata/calculations/getClassMetadata.js';
import { inject } from './metadata/decorators/inject.js';
import { injectable } from './metadata/decorators/injectable.js';
import { injectFromBase } from './metadata/decorators/injectFromBase.js';
import { injectFromHierarchy } from './metadata/decorators/injectFromHierarchy.js';
import { multiInject } from './metadata/decorators/multiInject.js';
import { named } from './metadata/decorators/named.js';
import { optional } from './metadata/decorators/optional.js';
import { postConstruct } from './metadata/decorators/postConstruct.js';
import { preDestroy } from './metadata/decorators/preDestroy.js';
import { tagged } from './metadata/decorators/tagged.js';
import { unmanaged } from './metadata/decorators/unmanaged.js';
import { ClassElementMetadataKind } from './metadata/models/ClassElementMetadataKind.js';
import { ResolvedValueElementMetadataKind } from './metadata/models/ResolvedValueElementMetadataKind.js';
import { plan } from './planning/actions/plan.js';
import { CacheBindingInvalidationKind } from './planning/models/CacheBindingInvalidationKind.js';
import { LazyPlanServiceNode } from './planning/models/LazyPlanServiceNode.js';
import { PlanResultCacheService } from './planning/services/PlanResultCacheService.js';
import { resolve } from './resolution/actions/resolve.js';
import { resolveBindingsDeactivations } from './resolution/actions/resolveBindingsDeactivations.js';
import { resolveModuleDeactivations } from './resolution/actions/resolveModuleDeactivations.js';
import { resolveServiceDeactivations } from './resolution/actions/resolveServiceDeactivations.js';
export { ActivationsService, bindingScopeValues, BindingService, bindingTypeValues, CacheBindingInvalidationKind, ClassElementMetadataKind, DeactivationsService, decorate, getBindingId, getClassMetadata, inject, injectable, injectFromBase, injectFromHierarchy, InversifyCoreError, InversifyCoreErrorKind, LazyPlanServiceNode, multiInject, named, optional, plan, PlanResultCacheService, postConstruct, preDestroy, resolve, resolveBindingsDeactivations, ResolvedValueElementMetadataKind, resolveModuleDeactivations, resolveServiceDeactivations, tagged, unmanaged, };
